src/Razor/test/Microsoft.AspNetCore.Razor.LanguageServer.Test/ProjectConfigurationStateSynchronizerTest.cs

```diff
@@ -252,17 +267,85 @@ public void ProjectConfigurationFileChanged_Changed_UntrackedProject_Noops()
+        {
+            // Arrange
+            var handle = new FullProjectSnapshotHandle(
+            "path/to/project.csproj",
+                RazorConfiguration.Default,
+                rootNamespace: "TestRootNamespace",
+                new ProjectWorkspaceState(Array.Empty<TagHelperDescriptor>(), LanguageVersion.CSharp5),
+                Array.Empty<DocumentSnapshotHandle>());
+
+            var filePath = "path/to/project.csproj";
+            var projectService = new Mock<RazorProjectService>(MockBehavior.Strict);
+            projectService.Setup(service => service.AddProject(handle.FilePath)).Verifiable();
+            projectService.Setup(p => p.UpdateProject(
+                filePath,
+                It.IsAny<RazorConfiguration>(),
+                It.IsAny<string>(),
+                It.IsAny<ProjectWorkspaceState>(),
+                It.IsAny<IReadOnlyList<DocumentSnapshotHandle>>()));
+
+            var synchronizer = GetSynchronizer(projectService.Object);
+            var changedDeserializer = CreateJsonFileDeserializer(handle);
+            var removedArgs = new ProjectConfigurationFileChangeEventArgs(filePath, RazorFileChangeKind.Removed, changedDeserializer);
+            var addedArgs = new ProjectConfigurationFileChangeEventArgs(filePath, RazorFileChangeKind.Added, changedDeserializer);
+            var changedArgs = new ProjectConfigurationFileChangeEventArgs(filePath, RazorFileChangeKind.Changed, changedDeserializer);
+
+            // Act
+            synchronizer.ProjectConfigurationFileChanged(addedArgs);
+            synchronizer.ProjectConfigurationFileChanged(changedArgs);
+            WaitForEnqueue(synchronizer).Wait();
 
             // Assert
+            projectService.Verify(p => p.UpdateProject(
+                filePath,
+                It.IsAny<RazorConfiguration>(),
+                It.IsAny<string>(),
+                It.IsAny<ProjectWorkspaceState>(),
+                It.IsAny<IReadOnlyList<DocumentSnapshotHandle>>()), Times.Once);
+
             projectService.VerifyAll();
         }
 
+        private async Task WaitForEnqueue(ProjectConfigurationStateSynchronizer synchronizer, bool hasTask = true)
+        {
+            if (hasTask)
+            {
+                var kvp = Assert.Single(synchronizer._projectInfoMap);
+                await Task.Factory.StartNew(() =>
+                {
+                    kvp.Value.ProjectUpdateTask.Wait();
+                }, CancellationToken.None, TaskCreationOptions.None, Dispatcher.ForegroundScheduler);
+            }
+            else
+            {
+                Assert.Empty(synchronizer._projectInfoMap);
+            }
+        }
+
+        private ProjectConfigurationStateSynchronizer GetSynchronizer(RazorProjectService razorProjectService)
+        {
+            var synchronizer = new ProjectConfigurationStateSynchronizer(Dispatcher, razorProjectService, FilePathNormalizer);
+            synchronizer.EnqueueDelay = 25;
```

Comment:
Could probably even bring this down to 1ms given you have more control now 😄

Reply:
1ms seems like it might be TOO fast, if there's GC or something we might execute both tasks when we expected to only run the last one. Maybe 5ms is a bit safer?